### PR TITLE
Laurie/ros2visualizers

### PR DESCRIPTION
**Pull request by LaurieCheers-unity**

### CODE REVIEW

File: com.unity.robotics.ros-tcp-connector/Runtime/TcpConnector/TopicVisualizationState.cs [hunk 1]
```diff
@@ -99,11 +97,14 @@ public void SetMessage(Message message, MessageMetadata meta)
...
             if (m_DrawingUpdatedAtFrameIndex != meta.FrameIndex)
             {
-                m_Contents.ClearVisual();
                 if (ShowDrawing)
                 {
                     m_Contents.ShowDrawing(true);
                 }
+                else
+                {
+                    m_Contents.ClearVisual();
```
Comment (LaurieCheers-unity) [inline com.unity.robotics.ros-tcp-connector/Runtime/TcpConnector/TopicVisualizationState.cs:106]: Don't call ClearVisual if we're doing ShowDrawing. This allows the recycling to work.

File: com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        SettingsBasedVisualFactory<TMessageType, TVisualizerSettings> visualizer = (SettingsBasedVisualFactory<TMessageType, TVisualizerSettings>)target;
+        visualizer.Topic = EditorGUILayout.TextField("Topic", visualizer.Topic);
+
+        m_Config = visualizer.Settings;
+        GUI.changed = false;
+        m_Config = (TVisualizerSettings)EditorGUILayout.ObjectField("Visualizer settings", m_Config, typeof(TVisualizerSettings), false);
+        visualizer.Settings = m_Config;
+
+        EditorGUI.indentLevel++;
```
Comment (at669) [inline com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs:25]: ah... i did not know about `GUI.changed` or `EditorGUI.indentLevel`... i see :') good to know!

File: com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        OnInspectorGUIForSettings(visualizer);
+        EditorGUI.indentLevel--;
+
+        if (GUI.changed)
+        {
+            visualizer.Redraw();
+        }
+    }
+
+    protected virtual void OnInspectorGUIForSettings(SettingsBasedVisualFactory<TMessageType, TVisualizerSettings> visualizer)
```
Comment (at669) [inline com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs:35]: just for my own understanding; this is if the target for this editor is null, it makes/caches a new one for whatever matches its custom editor attribute? is this the case of creating at runtime, or when would this happen?

File: com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/LaserScanVisualizerSettings.cs [hunk 4]
```diff
@@ -1,11 +1,76 @@
...
-[CreateAssetMenu(fileName = "LaserScanVisualizerSettings", menuName = "MessageVisualizations/Sensor/LaserScan", order = 1)]
-public class LaserScanVisualizerSettings : ScriptableObject
+[CreateAssetMenu(fileName = "LaserScanVisualizerSettings", menuName = "Robotics/Sensor Messages/LaserScan", order = 1)]
+public class LaserScanVisualizerSettings : VisualizerSettings<LaserScanMsg>
 {
-    public string topic;
-    public bool useIntensitySize;
-    public float pointRadius = 0.05f;
-    public float[] sizeRange = { 0, 100 };
+    public TFTrackingType m_TFTrackingType = TFTrackingType.Exact;
```
Comment (at669) [inline com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/LaserScanVisualizerSettings.cs:10]: nit/question: pascal case for public variables?

File: com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/PointCloud2VisualizerSettings.cs [hunk 5]
```diff
@@ -1,30 +1,140 @@
...
+
+            if (m_UseSizeChannel)
+            {
+                var size = BitConverter.ToSingle(message.data, iPointStep + sizeChannelOffset);
+                radius = Mathf.InverseLerp(m_SizeRange[0], m_SizeRange[1], size);
+            }
+
+            pointCloud.AddPoint(worldPoint, color, radius);
+        }
+        //pointCloud.Bake();
```
Comment (at669) [inline com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/PointCloud2VisualizerSettings.cs:115]: i forget--is there a reason this is commented out?

File: com.unity.robotics.message-visualizations/Runtime/Scripts/DrawingVisualFactory.cs [hunk 6]
```diff
@@ -4,17 +4,18 @@
...
     {
         [SerializeField]
-        protected string m_Topic;
+        string m_Topic;
+        public string Topic { get => m_Topic; set => m_Topic = value; }
 
         public virtual void Start()
         {
-            if (m_Topic == "")
+            if (m_Topic == null || m_Topic == "")
```
Comment (at669) [inline com.unity.robotics.message-visualizations/Runtime/Scripts/DrawingVisualFactory.cs:16]: can this be [`IsNullOrEmpty`](https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorempty?view=net-5.0)

**Review by at669 — APPROVED** [review]
most comments are just questions for my understanding, and just wondering about the URDF-Importer branch/version being used in the project you're testing in

File: com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/LaserScanVisualizerSettings.cs (hunk 4 above)
Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/LaserScanVisualizerSettings.cs:10]: Oh, good point, public variables definitely shouldn't have the m_ prefix.

File: com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/PointCloud2VisualizerSettings.cs (hunk 5 above)
Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/ScriptableObjects/PointCloud2VisualizerSettings.cs:115]: Should just remove this line I guess.
The intent of the Bake function was to free up memory: it generates the point-cloud mesh immediately and discards all the intermediate data. However, in practice we actually want to keep the intermediate data around so we can reuse it, and avoid garbage collection. Possibly we'll find a use for Bake later.

File: com.unity.robotics.message-visualizations/Runtime/Scripts/DrawingVisualFactory.cs (hunk 6 above)
Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Runtime/Scripts/DrawingVisualFactory.cs:16]: Yes.

File: com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs (hunk 3 above)
Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs:35]: Not sure what you mean... is this comment attached to the wrong line?

File: com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/DefaultVisualizerMultiDOFJointState.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 using System;
 using RosMessageTypes.Sensor;
-using Unity.Robotics.UrdfImporter;
+using RosSharp.Urdf;
```
[the remark this answers is not in the document]
Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Runtime/DefaultVisualizers/Sensor/DefaultVisualizerMultiDOFJointState.cs:3]: Ah, you're probably right... will try again. I was having trouble importing the latest URDF importer at the time, so I changed these instead.

File: com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs (hunk 3 above)
Comment (at669) [reply com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs:35]: oh, yes this was meant to go down to 39--my question really is more about what `CreateCachedEditor` does and in what case is `m_Editor` null?

Comment (LaurieCheers-unity) [reply com.unity.robotics.message-visualizations/Editor/SettingsBasedVisualizerEditor.cs:35]: CreateCachedEditor creates the editor for the scriptableobject, which we then immediately use to edit the object in this inspector window.
m_Editor is null the first time it runs this code. Actually I suspect the m_Editor == null check is wrong, we should be rerunning CreateCachedEditor if m_Config changes too...

**Pull request merged**